### ISSUE TITLE
RFE: update the syscall list to the latest Linux -rcX release for libseccomp v2.5

### ISSUE THREAD

**Issue by pcmoore** [seccomp/libseccomp#164]
We want to make sure we support the latest kernel release.

**Comment by pcmoore** [comment]
I'll go ahead and take this one.

**Comment by drakenclimber** [comment]
I ran `arch-syscall-validate` against the head of Linus' kernel branch, and it reports that libseccomp is fully up to date on the v5.7.0-rc6 syscalls.

But I'm still a little nervous that we might have missed something since we made some major changes here - both the *.csv file as well as the gperf changes.  @pcmoore do you have any thoughts on how we can reduce this "risk"?  Going through each syscall in each architecture sounds mind-numbingly boring - and error prone.

**Comment by pcmoore** [comment]
> I ran arch-syscall-validate against the head of Linus' kernel branch, and it reports that libseccomp is fully up to date on the v5.7.0-rc6 syscalls.

Thanks for checking this.  I've got a personal TODO item to setup a job which runs this regularly against Linus' tree and reports back but needless to say I haven't had time to do that yet.

I would vote to leave this issue open right up until we do the v2.5.0 release, mostly as a reminder to check this one last time.  Are you okay with that @drakenclimber?

> But I'm still a little nervous that we might have missed something since we made some major changes here - both the *.csv file as well as the gperf changes. @pcmoore do you have any thoughts on how we can reduce this "risk"? Going through each syscall in each architecture sounds mind-numbingly boring - and error prone.

Agreed on both the worry and the desire to do this check manually.

There are probably other ways to do this, but one possible solution would be to take the latest v2.4.x release and write a one-off tool which dumps it's internal syscall table into the same CSV formatted syscall table we use in the master branch.  At that point diff, or your favorite text comparison tool, could be used to inspect the tables and verify any differences.

I imagine you could steal/munge the code in `src/arch-syscall-dump.c` to get you most of the way there.

**Comment by drakenclimber** [comment]
> I would vote to leave this issue open right up until we do the v2.5.0 release, mostly as a reminder to check this one last time. Are you okay with that @drakenclimber?

Yes, definitely.  I had a couple free minutes a week or so ago and wanted to see the status of this issue.  Thankfully, things are looking pretty good.

> There are probably other ways to do this, but one possible solution would be to take the latest v2.4.x release and write a one-off tool which dumps it's internal syscall table into the same CSV formatted syscall table we use in the master branch. At that point diff, or your favorite text comparison tool, could be used to inspect the tables and verify any differences.
> 
> I imagine you could steal/munge the code in `src/arch-syscall-dump.c` to get you most of the way there.

Good idea.  I'll try to hack something together.  Thanks.

**Comment by drakenclimber** [comment]
> Good idea. I'll try to hack something together. Thanks.

[Here's](https://gist.github.com/drakenclimber/5ce9cf25d38f1742823af9060db14869) a script that uses arch-syscall-dump to build up a csv.  I have ironed out most of the quirks (e.g. mips using an offset of 4000 for its syscalls), but there are still several syscalls that are being flagged.  Either we'll have to improve the script :) - or check them by hand :(.

I have [uploaded the *.csv it generated](https://gist.github.com/drakenclimber/de2e836036622c96e66de4f926d1d3bf) for the release-2.4 branch.  Note that you will have to delete the RISCV column from the mainline csv since RISCV isn't supported in v2.4.

**Comment by drakenclimber** [comment]
Oh, and we should consider making some small tweaks to the [mainline csv](https://github.com/seccomp/libseccomp/blob/master/src/syscalls.csv).  Github can now display csv files if you don't confuse it too much.  Because of the comma in the comment line, `# kernel: 5.6.0 (Mon, 11 May 2020 04:28:57 +0000)`, github refuses to display the file as a csv.

**Comment by pcmoore** [comment]
> I have uploaded the *.csv it generated for the release-2.4 branch. Note that you will have to delete the RISCV column from the mainline csv since RISCV isn't supported in v2.4.

Don't leave me hanging here, how did it compare? :)

> ... Github can now display csv files if you don't confuse it too much ...

Oooh, that's nice!  Look at the [GH CSV docs](https://help.github.com/en/github/managing-files-in-a-repository/rendering-csv-and-tsv-data) I'm wondering if we shouldn't remove that header almost entirely and stick to a single header line?  Although I do like having the kernel version and date in there (and not just the git log), maybe something like this for row one?

```
#syscall (v5.6.0 2020-05-11),x86,x86_64,x32,arm,aarch64,...
```

**Comment by drakenclimber** [comment]
> Don't leave me hanging here, how did it compare? :)

```
$ diff -u release-2.4.csv mainline-syscalls.csv | diffstat
 mainline-syscalls.csv |  152 ++++++++++++++++++++++++++++++-----------------------------
 1 file changed, 79 insertions(+), 73 deletions(-)
```

**Comment by drakenclimber** [comment]
> Don't leave me hanging here, how did it compare? :)

I would estimate 80% of the syscalls match exactly.  A few percent aren't matching because of the parisc syscalls not matching (largely due to PNR), and several mismatches because my alphabetic sorting isn't matching the sorting of the mainline csv.

**Comment by drakenclimber** [comment]
> Oooh, that's nice! Looking at the [GH CSV docs](https://help.github.com/en/github/managing-files-in-a-repository/rendering-csv-and-tsv-data) I'm wondering if we shouldn't remove that header almost entirely and stick to a single header line? Although I do like having the kernel version and date in there (and not just the git log), maybe something like this for row one?
> 
> ```
> #syscall (v5.6.0 2020-05-11),x86,x86_64,x32,arm,aarch64,...
> ```

That works for me.  And if we are good citizens, we can make sure to list the kernel info in the git commit as well.  I think that should be sufficient.

**Comment by drakenclimber** [comment]
I dug further into the syscall differences between v2.4.3 and the head of master.

- First, I re-sorted the syscalls to remove those inconsistencies.  This removed ~20 or so diffs.
- Then I went line by line.  All of the remaining diffs (~45 or so) are due to PNR differences in the parisc or parisc64 lines.  Since you built the syscalls.csv from Linus' latest tree, I think the head of master is better than the v2.4 branch.  Is there a parisc person that we could ping to verify this?
- Finally, the `sync_file_range2` syscall changed from 341 to PNR on aarch64.  I have no clue if this is correct or not.  Almost all other architectures use PNR.

tl;dr - I think the syscall.csv in master is good to go.  There is a slight risk that parisc/parsic64 may have changed adversely.

**Comment by pcmoore** [comment]
Great, thank you @drakenclimber.  I know it isn't the most interesting thing to do, but it is *very* important.

> Is there a parisc person that we could ping to verify this?

Looking at the git log for "src/arch-parisc.c" I only see one name, maybe try sending Helge some email?  If push comes to shove, parisc/parisc64 are niche ABIs and if we can't verify things "soonish" I'm okay moving ahead with the release.

> Finally, the sync_file_range2 syscall changed from 341 to PNR on aarch64 ...

This concerns me slightly, we should follow up on that to see what the correct value is for aarch64.  One of the branches is clearly wrong (master or release-2.4), we need to figure out which one.

**Comment by drakenclimber** [comment]
> I checked the diff. The current CSV list is correct. When the parisc syscall list was converted to be parseable by scripts in kernel v5.0, those syscalls which were implemented as ni_syscall got removed. This affects e.g. afs_syscall, create_module, get_thread_area and others. Some syscalls are new, e.g. fsmount, move_mount and others. And all others (*time64) were added to cope with y2038 on 32bit platforms. So, overall, the seccomp syscall list for parisc and parisc64 are good.

Awesome!  Thanks so much for the help @hdeller!

I'll backport the parisc syscall changes to v2.4 and CC you on the patchset if you're okay with that.
